force/src/main/java/com/google/cdap/SalesforceToGCSAction.java

```diff
@@ -138,17 +142,37 @@ public void run(ActionContext actionContext) throws Exception {
                                                                          config.getClientSecret(), config.getUsername(),
                                                                          config.getPassword(), config.getApiVersion());
     JobInfo job = SalesforceBulkAPIs.createJob(config.getObject(), bulkConnection);
-    List<String> results = SalesforceBulkAPIs.runBulkQuery(config.query, bulkConnection, job);
-    for (String result : results) {
-      write(result);
+    runBulkQuery(config.query, bulkConnection, job);
+  }
+
+
+  private void runBulkQuery(String query, BulkConnection bulkConnection,
+                            JobInfo job) throws IOException, AsyncApiException, GeneralSecurityException {
+    BatchInfo info;
+    try (ByteArrayInputStream bout = new ByteArrayInputStream(query.getBytes())) {
+      info = bulkConnection.createBatchFromStream(job, bout);
+      for (int i = 0; i < 10000; i++) {
+        info = bulkConnection.getBatchInfo(job.getId(), info.getId());
+        if (BatchStateEnum.Completed == info.getState()) {
+          QueryResultList list = bulkConnection.getQueryResultList(job.getId(), info.getId());
+          for (String result : list.getResult()) {
+            write(bulkConnection.getQueryResultStream(job.getId(), info.getId(), result));
+          }
+
+          break;
+        } else if (BatchStateEnum.Failed == info.getState()) {
+          LOG.error("Failed " + info);
+          break;
+        } else {
+          LOG.debug("Waiting " + info);
```

Comment:
This should sleep for some time. Right now it will spam the API, I'm surprised it's not hitting rate limits or anything.

Reply:
Added a 1 second sleep.